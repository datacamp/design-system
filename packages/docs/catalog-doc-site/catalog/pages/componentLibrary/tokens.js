import tokens from '@datacamp/waffles-tokens';
import { Page } from 'catalog';
import React from 'react';

import CustomHeader from '../../components/CustomHeader';

export default () => {
  return (
    <main>
      <CustomHeader
        section="Component Library"
        subSection="Tokens"
        description="The primitive values used to style anything and everything"
      />

      <Page>
        <section className="dc-u-maxw-100pc">
          <div className="dc-card dc-u-p-24 dc-u-mt-8">
            <h4>About</h4>
            <p>
              The <code>@datacamp/waffles-tokens</code> package exposes a json
              file containing all the primitive values necessary to produce apps
              with the datacamp styling rules. These values can be accessed and
              used to style applications using a variety of tools, including
              CSS-in-JS libraries such as emotion.
            </p>
            <h4>Usage</h4>
            <p>
              To use any of the styles, simply import the{' '}
              <code>@datacamp/waffles-tokens</code> package and directly access
              the necessary properties. For example if the primary color is
              needed then it should be accessed as follows.
            </p>
            <pre>
              <code>{`import { colors } from '@datacamp/waffles-tokens';

const primaryColorValue = colors.primary;`}</code>
            </pre>
            <p>
              This hex code can then be used wherever is necessary. All the
              tokens can be accessed and used in the same way since it is just a
              json file.
            </p>
            <h4>Available tokens</h4>
            <pre>
              <code>{JSON.stringify(tokens, null, 2)}</code>
            </pre>
          </div>
        </section>
      </Page>
    </main>
  );
};
