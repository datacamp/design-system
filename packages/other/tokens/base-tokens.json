{
  "sizing": {
    "small": "24px",
    "medium": "36px",
    "large": "48px"
  },
  "spacing": {
    "xsmall": "4px",
    "small": "8px",
    "medium": "16px",
    "large": "24px",
    "xlarge": "32px"
  },
  "colors": {
    "primary": {
      "navyLight": "#213147",
      "navy": "#05192D",
      "navyDark": "#000820",
      "navySubtleTextOnDark": "#9BA3AB",
      "navySubtleTextOnLight": "#65707C",
      "greenLight": "#65FF8F",
      "green": "#03EF62",
      "greenDark": "#00C53B",
      "greenDarkText": "#008700"
    },
    "neutral": {
      "white": "#FFFFFF",
      "greySubtle": "#F7F7FC",
      "greyLight": "#EFEFEF",
      "grey": "#E8E8EA",
      "greyMedium": "#D9D9E2",
      "greyDark": "#848C92",
      "beigeSubtle": "#FFFBF3",
      "beigeLight": "#F7F3EB",
      "beige": "#EFEBE4",
      "beigeMedium": "#E5E1DA"
    },
    "accent": {
      "redLight": "#FF782D",
      "red": "#FF5400",
      "redDark": "#DD3400",
      "redDarkText": "#C01100",
      "orangeLight": "#FFBC4B",
      "orange": "#FF931E",
      "orangeDark": "#D87300",
      "orangeDarkText": "#B75900",
      "yellowLight": "#FFEC3C",
      "yellow": "#FCCE0D",
      "yellowDark": "#CFA600",
      "yellowDarkText": "#907000",
      "blueLight": "#60E7FF",
      "blue": "#06BDFC",
      "blueDark": "#009BD8",
      "blueDarkText": "#0075AD",
      "purpleLight": "#974DFF",
      "purple": "#7933FF",
      "purpleDark": "#5646A5",
      "purpleDarkText": "#5646A5",
      "pinkLight": "#FF95CF",
      "pink": "#FF6EA9",
      "pinkDark": "#DC4D8B",
      "pinkDarkText": "#BF3072"
    },
    "brand": {
      "brandFacebook": "#1778F2",
      "brandTwitter": "#00ACEE",
      "brandGoogle": "#DB4437",
      "brandLinkedIn": "#0E76A8"
    }
  },
  "borderRadius": {
    "medium": "4px",
    "circle": "50%"
  },
  "borderWidth": {
    "thin": "1px",
    "medium": "2px",
    "thick": "3px",
    "xthick": "4px"
  },
  "boxShadow": {
    "thin": {
      "value": {
        "x": "0px",
        "y": "0px",
        "spread": "0px",
        "color": "rgba(5, 25, 45, 0.3)",
        "blur": "2px"
      }
    },
    "medium": {
      "value": {
        "x": "0px",
        "y": "1px",
        "spread": "-1px",
        "color": "rgba(5, 25, 45, 0.3)",
        "blur": "4px"
      }
    },
    "thick": {
      "value": {
        "x": "0px",
        "y": "3px",
        "spread": "-1px",
        "color": "rgba(5, 25, 45, 0.3)",
        "blur": "5px"
      }
    },
    "xthick": {
      "value": {
        "x": "0px",
        "y": "8px",
        "spread": "-4px",
        "color": "rgba(5, 25, 45, 0.3)",
        "blur": "12px"
      }
    }
  },
  "fontWeights": {
    "regular": "Regular",
    "bold": "Bold"
  },
  "fontFamilies": {
    "sansSerif": "Studio Feixen Sans",
    "mono": "JetBrains Mono NL"
  },
  "fontSizes": {
    "small": "12px",
    "medium": "14px",
    "large": "16px",
    "xlarge": "18px",
    "xxlarge": "20px",
    "huge": "28px"
  },
  "letterSpacing": {
    "tight": "-0.5px",
    "default": "0px",
    "relaxed": "1.5px"
  },
  "lineHeights": {
    "tight": "100%",
    "default": "125%",
    "relaxed": "150%"
  },
  "opacity": {
    "low": "15%",
    "medium": "40%",
    "high": "60%"
  },
  "zIndex": {
    "default": 1,
    "sticky": 100,
    "dropdown": 7000,
    "overlay": 8000,
    "modal": 9000,
    "toast": 10000
  },
  "breakpoints": {
    "small": "480px",
    "medium": "820px",
    "large": "1480px"
  },
  "paragraphSpacing": {
    "default": "0px"
  },
  "typography": {
    "heading": {
      "h1": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.default",
        "fontSize": "$fontSizes.xxlarge",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "h2": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.default",
        "fontSize": "$fontSizes.xlarge",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "h3": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.default",
        "fontSize": "$fontSizes.large",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "h4": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.default",
        "fontSize": "$fontSizes.medium",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "display": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.default",
        "fontSize": "$fontSizes.huge",
        "letterSpacing": "$letterSpacing.tight",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "chapeau": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.default",
        "fontSize": "$fontSizes.small",
        "letterSpacing": "$letterSpacing.relaxed",
        "paragraphSpacing": "$paragraphSpacing.default"
      }
    },
    "text": {
      "body": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.regular",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.medium",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "bodyBold": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.medium",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "bodySmall": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.regular",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.small",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "bodySmallBold": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.small",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "bodyLarge": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.regular",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.large",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "bodyLargeBold": {
        "fontFamily": "$fontFamilies.sansSerif",
        "fontWeight": "$fontWeights.bold",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.large",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "code": {
        "fontFamily": "$fontFamilies.mono",
        "fontWeight": "$fontWeights.regular",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.small",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      },
      "codeLarge": {
        "fontFamily": "$fontFamilies.mono",
        "fontWeight": "$fontWeights.regular",
        "lineHeight": "$lineHeights.relaxed",
        "fontSize": "$fontSizes.medium",
        "letterSpacing": "$letterSpacing.default",
        "paragraphSpacing": "$paragraphSpacing.default"
      }
    }
  }
}
