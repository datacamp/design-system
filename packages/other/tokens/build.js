const styleDictionary = require('style-dictionary');

const coreJsonFormatter = require('./formats/coreJson');
const jsonFormatter = require('./formats/jsonFormatter');
const scssVariableFormatter = require('./formats/scssVariableFormatter');
const gradientTransform = require('./transforms/gradient');

const transforms = ['attribute/cti', 'attribute/color', 'waffles/gradient'];

const tokenBuilder = styleDictionary.extend({
  platforms: {
    'core-json': {
      buildPath: 'lib/',
      files: [{ destination: 'core-tokens.json', format: 'waffles/core-json' }],
      transforms,
    },
    'full-json': {
      buildPath: 'lib/',
      files: [{ destination: 'future-tokens.json', format: 'waffles/json' }],
      transforms,
    },
    scss: {
      buildPath: 'lib/',
      files: [
        { destination: 'variables.scss', format: 'waffles/scss-variables' },
      ],
      transforms,
    },
  },
  source: ['properties/**/*.json'],
});

tokenBuilder.registerTransform({
  matcher: (prop) => prop.attributes.category === 'gradient',
  name: 'waffles/gradient',
  transformer: gradientTransform,
  type: 'value',
});

tokenBuilder.registerFormat({
  formatter: coreJsonFormatter,
  name: 'waffles/core-json',
});

tokenBuilder.registerFormat({
  formatter: jsonFormatter,
  name: 'waffles/json',
});

tokenBuilder.registerFormat({
  formatter: scssVariableFormatter,
  name: 'waffles/scss-variables',
});

tokenBuilder.buildAllPlatforms();
