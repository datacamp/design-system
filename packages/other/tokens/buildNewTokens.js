// Generate new design tokens JS file based on JSON file with Figma Tokens definitions
// It will be used to generate both ESM and CJS modules, and TS typings

// fontWeights, fontFamilies, and boxShadow have non-standard values to work with Figma Tokens
// typography and paragraphSpacing are defined for convenience  and are removed
// colors are grouped for convenience and are flattened

const fs = require('fs');
const util = require('util');

const baseTokens = require('./base-tokens.json'); // Already parsed to JS

// Mappign for Figma Tokens values and CSS couterparts

const fontWeights = {
  Bold: 800,
  Regular: 400,
};

const fontFamilies = {
  'JetBrains Mono NL':
    "JetBrainsMonoNL, Menlo, Monaco, 'Courier New', monospace",
  'Studio Feixen Sans': 'Studio-Feixen-Sans, Arial, sans-serif',
};

// Transformations for colors, boxShadow, fontWeights, fontFamilies, and lineHeights

function transformedColors(baseColors) {
  return Object.entries(baseColors).reduce((flattenedColors, currentEntry) => {
    const groupedColors = currentEntry[1];
    return Object.assign(flattenedColors, { ...groupedColors });
  }, {});
}

function transformedBoxShadows(baseBoxShadows) {
  return Object.fromEntries(
    Object.entries(baseBoxShadows).map((entry) => {
      const [key, baseBoxShadow] = entry;
      const { blur, color, spread, x, y } = baseBoxShadow.value;
      // Regular CSS box-shadow value
      const regularBoxShadow = `${x} ${y} ${blur} ${spread} ${color}`;

      return [key, regularBoxShadow];
    }),
  );
}

function transformedFontWeights(baseFontWeights) {
  return Object.fromEntries(
    Object.entries(baseFontWeights).map((entry) => {
      const [key, baseFontWeight] = entry;
      return [key, fontWeights[baseFontWeight]];
    }),
  );
}

function transformedFontFamilies(baseFontFamilies) {
  return Object.fromEntries(
    Object.entries(baseFontFamilies).map((entry) => {
      const [key, baseFontFamily] = entry;
      return [key, fontFamilies[baseFontFamily]];
    }),
  );
}

function transformedLineHeights(baseLineHeights) {
  return Object.fromEntries(
    Object.entries(baseLineHeights).map((entry) => {
      const [key, baseLineHeight] = entry;
      // Because percanteges are really quirky in CSS, convert them to unitless
      const regularLineHeight = parseFloat(baseLineHeight) / 100;
      return [key, regularLineHeight];
    }),
  );
}

// Apply trasformations and remove typography and paragraphSpacing sections
function transformedBaseTokens(tokens) {
  const transformedTokens = {
    ...tokens,
    boxShadow: {
      ...transformedBoxShadows(tokens.boxShadow),
    },
    colors: {
      ...transformedColors(tokens.colors),
    },
    fontFamilies: {
      ...transformedFontFamilies(tokens.fontFamilies),
    },
    fontWeights: {
      ...transformedFontWeights(tokens.fontWeights),
    },
    lineHeights: {
      ...transformedLineHeights(tokens.lineHeights),
    },
  };

  delete transformedTokens.typography;
  delete transformedTokens.paragraphSpacing;

  return transformedTokens;
}

// Create separate export for each token group (like colors, lineHeights, etc.)
function exportForEachTokenGroup(tokens) {
  let exports = '';
  Object.entries(tokens).forEach((entry) => {
    const [key, tokenGroupValues] = entry;
    // Stringify values in a token group
    const stringifiedValues = util.inspect(tokenGroupValues, { depth: null });
    exports += `export const ${key} = ${stringifiedValues};\n`;
  });

  return exports;
}

// To make it more ergonomic create default export
function tokensDefaultExport(tokens) {
  // Comma separated list of exports
  let groupsToExport = '';
  const tokensKeys = Object.keys(tokens);

  tokensKeys.forEach((key, index) => {
    let separator = '';
    if (index < tokensKeys.length - 1) {
      separator = ', ';
    }
    groupsToExport += `${key}${separator}`;
  });

  return `export default { ${groupsToExport} };`;
}

// Content of ES modules, later written to file
// Serves as a base for generating ESM and CJS modules with Babel
function tokensToEsModule(tokens) {
  let moduleContent = exportForEachTokenGroup(tokens);
  moduleContent += tokensDefaultExport(tokens);

  return moduleContent;
}

const newTokens = tokensToEsModule(transformedBaseTokens(baseTokens));

// Assuming lib directory already exists
fs.writeFileSync('./lib/tokens.js', newTokens);

// Exporting utils for unit tests
module.exports = {
  exportForEachTokenGroup,
  tokensDefaultExport,
  tokensToEsModule,
  transformedBaseTokens,
  transformedBoxShadows,
  transformedColors,
  transformedFontFamilies,
  transformedFontWeights,
  transformedLineHeights,
};
