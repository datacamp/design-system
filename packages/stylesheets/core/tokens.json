{
  "spatial": {
    "multiplier": "2",
    "sizes": "2 4 8 12 16 24 32 36 48 64 80 96 128",
    "commonSizes": "8 12 16 24 32 36 48 64",
    "measure": "160 192 256 320 480 640 800 960",
    "percentages": "1 5 10 15 20 25 30 40 44 50 60 70 75 80 90 100",
    "angles": "30 45 60 90 180 270 360",
    "zIndex": "10 20 30 40 50 60 70 80 90 100 999"
  },
  "colors": {
    "blackTransparent": "'rgba(0, 0, 0, 0.2)'",
    "green": "#36d57d",
    "greenLight": "#a5ecc5",
    "geyser": "#cfdce1",
    "grey": "#4d5356",
    "greyDark": "#3d4251",
    "greyLight": "#d1d3d8",
    "greyLighter": "#e6eaeb",
    "greyOslo": "#859094",
    "orange": "#ff9400",
    "orangeDark": "#ad662c",
    "porcelain": "#f0f4f5",
    "primary": "#3ac",
    "primaryDark": "#263e63",
    "primaryDarker": "#14243e",
    "primaryDarkest": "#0c1626",
    "primaryLight": "#7ecce2",
    "primaryLighter": "#d5eaef",
    "primaryLightest": "#ebf4f7",
    "purple": "#8468c4",
    "purpleDark": "#5d488c",
    "purpleLight": "#a786f3",
    "raven": "#737a80",
    "red": "#fe5c5c",
    "redLight": "#ffb5b5",
    "secondary": "#ffc844",
    "secondaryLight": "#fbe28d",
    "white": "#fff",
    "whiteTransparent": "'rgba(255, 255, 255, 0.2)'"
  },
  "gradients": {
    "chambrayCloudburst": "linear-gradient(268deg, #405f8f, #1c3050)",
    "puertoricoDeyork": "linear-gradient(32deg, #37bbab, #75c97e)",
    "blueLight": "linear-gradient(-143deg, #7dd4e7, #7099d7 100%)",
    "grey": "linear-gradient(-135deg, #e7f1f6, #b6d1e5)",
    "primary": "linear-gradient(141deg, #2388b0, #3ac)",
    "primaryDark": "linear-gradient(141deg, #20748d, #164d62)",
    "primaryLight": "linear-gradient(180deg, rgba(255, 255, 255, 0), rgba(51, 170, 204, 0.075))",
    "purple": "linear-gradient(-136deg, #c06ac8, #8468c4)",
    "red": "linear-gradient(45deg, #f06a8c, #f6a38d)",
    "redLight": "linear-gradient(180deg, rgba(255, 255, 255, 0), rgba(254, 92, 92, 0.075))",
    "secondary": "linear-gradient(0deg, #ffc844, #fcd86c)",
    "turquoise": "linear-gradient(-135deg, #b8eef4, #78c8de)"
  },
  "fontFamily": {
    "sansSerif": "'Lato'",
    "mono": "'Roboto Mono'"
  },
  "fontSize": {
    "base": "16px",
    "micro": "0.75rem",
    "small": "0.875rem",
    "h1": "2.5rem",
    "h2": "2rem",
    "h3": "1.5rem",
    "h4": "1.25rem",
    "h5": "1.125rem",
    "h6": "1rem"
  },
  "fontWeight": {
    "bold": "700",
    "light": "300",
    "regular": "400"
  },
  "lineHeight": {
    "base": "1.5",
    "heading": "1.25"
  },
  "border": {
    "radius": "4px"
  },
  "boxShadow": {
    "color": "rgba(0, 0, 0, 0.15)",
    "border": "0 0 0 1px #e6eaeb",
    "sm": "0 1px 2px 0 rgba(0, 0, 0, 0.15)",
    "md": "0 1px 3px 0 rgba(0, 0, 0, 0.15)",
    "lg": "0 3px 5px 0 rgba(0, 0, 0, 0.15)",
    "xl": "0 4px 8px 0 rgba(0, 0, 0, 0.15)",
    "xxl": "0 10px 20px 0 rgba(0, 0, 0, 0.15)",
    "xlGreen": "0 4px 8px 0 rgba(54, 213, 125, 0.3)",
    "xlOrange": "0 4px 8px 0 rgba(255, 148, 0, 0.3)",
    "xlPurple": "0 4px 8px 0 rgba(132, 104, 196, 0.3)"
  },
  "animation": {
    "duration": "0.3s",
    "ease": "cubic-bezier(0.77, 0, 0.175, 1)",
    "transition": "0.3s cubic-bezier(0.77, 0, 0.175, 1)"
  },
  "breakpoints": {
    "xs": "480px",
    "sm": "768px",
    "md": "992px",
    "lg": "1200px",
    "xl": "1366px",
    "ws": "1680px",
    "4K": "1920px",
    "5K": "2560px",
    "belowXs": "479px",
    "belowSm": "767px",
    "belowMd": "991px",
    "belowLg": "1199px",
    "belowXl": "1365px",
    "belowWs": "1679px",
    "below4K": "1919px",
    "below5K": "2559px"
  }
}
