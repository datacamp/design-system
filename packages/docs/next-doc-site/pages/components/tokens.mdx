import Layout from '../../components/page-layout';
import Header from '../../components/page-header';
import Section from '../../components/page-section';
import FormattedTokens from '../../components/formatted-tokens';

export default Layout;

<Header
  pageTitle="Tokens"
  overview="The primitive values used to style anything and everything."
/>

<Section>

## Overview

The `@datacamp/waffles-tokens` package exposes all the primitive values necessary to style apps according to [Design Guidelines](https://www.figma.com/file/Flh00IKR70ff0bRvFW2Iou/%5BDC%5D-Design-Guidelines-Exploration?node-id=376%3A1144). These values can be accessed and used using a variety of tools, including CSS-in-JS libraries such as emotion.

### Usage

To use any of the styles, simply import the `@datacamp/waffles-tokens` package and directly access the necessary properties. For example if the green color is needed then it should be accessed as follows.

```
import tokens from '@datacamp/waffles-tokens';

const greenColor = tokens.colors.green;
```

This hex code can then be used wherever is necessary. All the tokens can be accessed and used in the same way.

### Colors

For each of the primary colours, there are 3 variations. The standard color e.g. `green` should be used as a priority, but there is also a dark and a light version of every colour. There is also a version to use as a text color when being used against a light coloured background, e.g. `greenDarkText`. For some colors this is the same as one of the other variants, but it will always have sufficient contrast against one of the lighter backgrounds. When working against a dark background, the standard colours are appropriate.

### Accessibility

Use `blue` (in some rare cases `blueDark` is allowed) color for interactive components focus styles. For disabled elements use `medium` opacity.

</Section>

<Section>

## Available Tokens

<FormattedTokens />

</Section>

<Section>

## Migration Guide

### Tokens (v1 -> v2)

Tokens have been updated to stay in line with [Design Guidelines](https://www.figma.com/file/Flh00IKR70ff0bRvFW2Iou/%5BDC%5D-Design-Guidelines-Exploration?node-id=376%3A1144) and to be compatible with [Figma Tokens](https://www.figma.com/community/plugin/843461159747178978/Figma-Tokens) plugin.

#### Deprecated tokens groups

- 游뚿 `spatial` group is deprecated
  - use tokens from `sizing`, `spacing`, and `zIndex` groups
- 游뚿 `border` group is deprecated
  - use tokens from `borderRadius` group
- 游뚿 `animation` group is deprecated

#### Completely new tokens groups

- 游릭 new group `sizing`
- 游릭 new group `spacing`
- 游릭 new group `borderWidth`
- 游릭 new group `opacity`
- 游릭 new group `zIndex`
- 游릭 new group `borderRadius`
  - 游뚿 `border.radius` -> `borderRadius.medium`
  - 游릭 new `circle` border radius token

#### Updated tokens groups

- 游댃 changes in `colors` group:
  - 游릭 new brand colors `brandFacebook`, `brandTwitter`, `brandGoogle`, `brandLinkedIn`
  - 游릭 new text colors `navySubtleTextOnDark`, `navySubtleTextOnLight`
  - 游릭 new `greyDark` color
  - 游뚿 `navyText` -> `navy`
  - 游뚿 `greenText` -> `greenDarkText`
  - 游뚿 `redText` -> `redDarkText`
  - 游뚿 `orangeText` -> `orangeDarkText`
  - 游뚿 `yellowText` -> `yellowDarkText`
  - 游뚿 `blueText` -> `blueDarkText`
  - 游뚿 `purpleText` -> `purpleDarkText`
  - 游뚿 `pinkText` -> `pinkDarkText`
  - 游뚿 `beige100` -> `beigeSubtle`
  - 游뚿 `beige200` -> `beigeLight`
  - 游뚿 `beige300` -> `beige`
  - 游뚿 `beige400` -> `beigeMedium`
  - 游뚿 `grey100` -> `greySubtle`
  - 游뚿 `grey200` -> `greyLight`
  - 游뚿 `grey300` -> `grey`
  - 游뚿 `grey400` -> `greyMedium`
- 游댃 changes in `breakpoints` group
  - 游릭 new `small`, `medium`, `large` breakpoints
  - 游뚿 all old tokens are deprecated
- 游댃 changes in `boxShadow` group
  - 游릭 new `thin`, `medium`, `thick`, `xthick` tokens
  - 游뚿 all old tokens are deprecated

#### Renamed and updated tokens groups

- 游댃 `fontFamily` -> `fontFamilies`
  - 游뚿 tokens now contains whole CSS font stacks, no need to specify additional fallback fonts
- 游댃 `fontSize` -> `fontSizes`
  - 游뚿 `micro` -> `small`
  - 游뚿 `small` -> `small`
  - 游뚿 `base` -> `medium`
  - 游뚿 `h6` -> `small`
  - 游뚿 `h5` -> `small`
  - 游뚿 `h4` -> `medium`
  - 游뚿 `h3` -> `large`
  - 游뚿 `h2` -> `xlarge`
  - 游뚿 `h1` -> `xxlarge`
  - 游릭 new `huge` token
- 游댃 `fontWeight` -> `fontWeights`
- 游댃 `letterSpacings` -> `letterSpacing`
  - 游뚿 `base` -> `default`
  - 游뚿 `mediumHeading` -> `tight`
  - 游뚿 `largeHeading` -> `tight`
  - 游릭 new `relaxed` token
- 游댃 `lineHeight` -> `lineHeights`
  - in most cases use the same `lineHeight` in tandem with `fontSize`
  - 游뚿 `base` -> `medium`
  - 游뚿 `heading` -> `large`
  - 游뚿 `largeHeading` -> `xlarge`
  - 游릭 new `small` token
  - 游릭 new `xxlarge` token

### Components

Most components feature only slight visual changes, without any updates to their props APIs (icons are the only exception).

#### Icons (v4 -> v5)

- 游뚿 `size` accepts `medium` (default), `small`, and `xsmall` values instead of numeric `24`, `18`, and `12`
  - in most cases, it is advised to use the default value
  - under the hood they map to 16, 14, and 12 numeric values

#### Button (v7 -> v8)

- font size for `small` and `medium` `Button` is set to `fontSizes.medium` (14px)
- adjusted icon size for `CloseButton` and social media buttons
- updated dropdown item elements dimensions for `UserAccountMenu`

#### Text (v5 -> v6)

- font size for `Small`, `Code`, and `CodeBlock` is set to `fontSizes.small` (12px)
- font size for `Paragraph`, `Link`, `Emphasis`, `Strong`, `Text`, and `ListItem` is set to `fontSizes.medium` (14px)
- font size, line height, and letter spacing are individually adjusted for each `Heading` size
- 丘멆잺 recommended combination of `size` for each HTML heading level for `Heading` component:
  - `h1` -> `700`
  - `h2` -> `650`
  - `h3` -> `600`
  - `h4` -> `500`
- font size for `small` `Badge` is set to `fontSizes.small` and for `large` one to `fontSizes.medium`

#### Toast (v3 -> v4)

- title font size is set to match the description
- position of icons are adjusted

#### Modals (v7 -> v8)

- reduce spacings around `Dialog.Header`, `Dialog.Body`, and `Dialog.Footer`

#### Tooltip (v0.2 -> v1)

- reduce spacings around the text

</Section>
