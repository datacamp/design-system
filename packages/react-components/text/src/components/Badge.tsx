import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';
import tinycolor from 'tinycolor2';

import Strong from './Strong';

interface BadgeProps {
  /**
   * The content to render in the Badge
   */
  children: string;
  /**
   * Additionaly css className to add to the rendered element
   */
  className?: string;
  /**
   * The colour of the Badge. This can be any css colour, but it is recommended
   * to use a colour from @datacamp/waffles-tokens. The text color will be set
   * automatically based on this colour.
   */
  color: string;
  /**
   * The size of the Badge to render.
   */
  size?: 'small' | 'large';
}

const baseStyle = css({
  borderRadius: 4,
  display: 'inline-block',
  textTransform: 'uppercase',
});

const sizeStyles = {
  large: {
    fontSize: tokens.size.font[300].value,
    lineHeight: '24px',
    paddingLeft: tokens.size.space[8].value,
    paddingRight: tokens.size.space[8].value,
  },
  small: {
    fontSize: tokens.size.font[200].value,
    lineHeight: '18px',
    paddingLeft: tokens.size.space[4].value,
    paddingRight: tokens.size.space[4].value,
  },
};

const getTextColor = (backgroundColor: string): string => {
  // This should be 4.5 for compliance. 2.7 is just low enough to give white
  // text on #3ac background
  if (tinycolor.readability(backgroundColor, 'white') > 2.7) {
    return 'white';
  }
  return 'rgba(12, 22, 38, 0.8)';
};

const Badge = ({
  children,
  className,
  color,
  size = 'small',
}: BadgeProps): React.ReactElement => (
  <Strong
    className={className}
    css={css(
      baseStyle,
      { backgroundColor: color, color: getTextColor(color) },
      sizeStyles[size],
    )}
  >
    {children}
  </Strong>
);

export default Badge;
