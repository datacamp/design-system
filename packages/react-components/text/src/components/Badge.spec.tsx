import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import React from 'react';

import Badge from './Badge';

describe('Badge', () => {
  const testContent = 'test content';
  it('renders with the provided colour, className and content', async () => {
    const { getByText } = await axeRender(
      <Badge className="test-class" color="black">
        {testContent}
      </Badge>,
    );
    const element = getByText(testContent) as HTMLElement;
    expect(element).toHaveStyle('background-color: black');
    expect(element).toHaveClass('test-class');
    expect(element).toMatchSnapshot();
  });

  describe('size', () => {
    it('renders a small badge', async () => {
      const { getByText } = await axeRender(
        <Badge color="black" size="small">
          {testContent}
        </Badge>,
      );
      const element = getByText(testContent) as HTMLElement;
      expect(element).toHaveStyle(`font-size: ${tokens.size.font[200].value}`);
      expect(element).toHaveStyle(`line-height: 18px`);
      expect(element).toHaveStyle(
        `padding-left: ${tokens.size.space[4].value}px`,
      );
      expect(element).toHaveStyle(
        `padding-right: ${tokens.size.space[4].value}px`,
      );
    });
    it('renders a large badge', async () => {
      const { getByText } = await axeRender(
        <Badge color="black" size="large">
          {testContent}
        </Badge>,
      );
      const element = getByText(testContent) as HTMLElement;
      expect(element).toHaveStyle(`font-size: ${tokens.size.font[300].value}`);
      expect(element).toHaveStyle(`line-height: 24px`);
      expect(element).toHaveStyle(
        `padding-left: ${tokens.size.space[8].value}px`,
      );
      expect(element).toHaveStyle(
        `padding-right: ${tokens.size.space[8].value}px`,
      );
    });
  });

  describe('textColor', () => {
    it('sets the text color to white on a primary background', async () => {
      const { getByText } = await axeRender(
        <Badge color={tokens.color.opaque.primary.value.hex}>
          {testContent}
        </Badge>,
      );
      const element = getByText(testContent) as HTMLElement;
      expect(element).toHaveStyle('color: white');
    });

    it('sets the text color to dark transparent grey on a light background', async () => {
      const { getByText } = await axeRender(
        <Badge color={tokens.color.opaque.greyLighter.value.hex}>
          {testContent}
        </Badge>,
      );
      const element = getByText(testContent) as HTMLElement;
      expect(element).toHaveStyle('color: rgba(12, 22, 38, 0.8)');
    });
  });
});
