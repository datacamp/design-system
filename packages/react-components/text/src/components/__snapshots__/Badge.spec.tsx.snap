// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders with the provided colour, className and content 1`] = `
.emotion-0 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-style: normal;
  font-weight: 800;
  line-height: 1.5;
  border-radius: 4px;
  display: inline-block;
  text-transform: uppercase;
  background-color: #05192D;
  color: #FFFFFF;
  font-size: 12px;
  line-height: 18px;
  padding-left: 4px;
  padding-right: 4px;
}

<strong
  class="test-class emotion-0"
>
  test content
</strong>
`;
